### ISSUE TITLE
Bug: [prefer-reduce-type-parameter] unfixable reporting

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3440]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/<rule>": ["prefer-reduce-type-parameter"]
  }
}
```

```TS
// first case
export function normalizeParams<P>(params: P): P {
  return Object.entries(params).reduce((acc, [key, value]) => {
    value = typeof value === 'string' ? value.trim() : value;
    if (value) {
      acc[key as keyof P] = value;
    }
    return acc;
  }, {} as P);
}

// second case
export type Keys<T> = { [Key in keyof T]: Key };

export const toKeys = <T>(source: T): Keys<T> =>
  Object.keys(source).reduce((acc, k) => {
    const key = k as keyof T;
    acc[key] = key;
    return acc;
  }, {} as Keys<T>);
```

**Expected Result**

Ignore such cases

**Actual Result**

Error reporting

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.24.0` |
| `@typescript-eslint/parser`        | `4.24.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.27.9` |
| `node`                             | `12.22.1` |

**Comment by bradzacher** [comment]
Both of your examples are ***completely unsound*** due to the `as` though.
You're *forcing* typescript to treat `P`/`T` as an object, but they are unbounded generics - meaning they can be non-objects.

```ts
const str1 = normalizeParams('foo');
// typeof str1 === 'foo'
// but at runtime
// str1 === {0: 'f', 1: 'o', 2: 'o'}

const num1 = normalizeParams(1);
// typeof num1 === 1
// but at runtime
// num1 === {}



const str2 = normalizeParams('foo');
// typeof str2 === string
// but at runtime
// str2 === {0: "0", 1: "1", 2: "2"}

const num1 = normalizeParams(1);
// typeof num1 === number
// but at runtime
// num1 === {}
```

Doing an `as` assertion really is unsafe here.
So the rule is really doing its job here by enforcing you don't write broken code.

**Comment by JounQin** [comment]
Then what about adding `extends object` after the generic?

```ts
export function normalizeParams<P extends object>(params: P): P {
  return Object.entries(params).reduce((acc, [key, value]) => {
    value = typeof value === 'string' ? value.trim() : value;
    if (value) {
      acc[key as keyof P] = value;
    }
    return acc;
  }, {} as P);
}
```

Or, do you have any recommendation to handle these cases?

**Comment by bradzacher** [comment]
It's hard to say exactly - TS does treat generics kind of weirds in conjunction with empty object types.

This all boils down to the `object` type being super hard to use at all.

**Comment by JounQin** [comment]
Actually I respectly disagree with

> This also all boils down to the `object` type being super hard to use at all.

It is very useful when you want a generic which must be an Object, not primitive nor Array, there is no alternative, the suggestions `Record` or `unknown` are not useful here. Of course this is another topic/rule question.

**Comment by JoshuaKGoldberg** [comment]
I actually just hit this in the typescript-eslint codebase 😄:

https://github.com/typescript-eslint/typescript-eslint/blob/9e35ef9af3ec51ab2dd49336699f3a94528bb4b1/packages/eslint-plugin/src/rules/naming-convention-utils/parse-options.ts#L82-L90

[Simplified in our playground](https://typescript-eslint.io/play#ts=4.9.3&sourceType=module&code=C4TwDgpgBACghgJwM4QCYHkzAJYHsB2SUAvFAEoQDGuCqAPAORwNQA+UDARgwDRT4BXALacICAHwBuAFDSAZgPyUcBKGEQpMKwgAoAlAC5YGtFryEoAb2lQo1QsCgIISAQBtHpANpNeHbgC6AHTOqAKUEDo6cJSUfADWeiTiVja2UDGUXvEBJFDxMulOEMACCPgZsYVQAL58ljUZRPDIpljmSHoyac6l5cWuHjI1QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4AmsMolpFStYgENoUgLaJ8vdFF7QA9tEjgwAXxC6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA):

```ts
type ParsedOptions = Record<'a' | 'b', number>;

function parseOptions(): ParsedOptions {
  const result = ['a', 'b'].reduce((acc, k) => {
    acc[k] = k;
    return acc;
  }, {} as ParsedOptions);

  return result;
}
```

**Comment by JoshuaKGoldberg** [comment]
I actually just hit this in the typescript-eslint codebase 😄:

https://github.com/typescript-eslint/typescript-eslint/blob/9e35ef9af3ec51ab2dd49336699f3a94528bb4b1/packages/eslint-plugin/src/rules/naming-convention-utils/parse-options.ts#L82-L90

[Simplified in our playground](https://typescript-eslint.io/play#ts=4.9.3&sourceType=module&code=C4TwDgpgBACghgJwM4QCYHkzAJYHsB2SUAvFAEoQDGuCqAPAORwNQA+UDARgwDRT4BXALacICAHwBuAFDSAZgPyUcBKGEQpMKwgAoAlAC5YGtFryEoAb2lQo1QsCgIISAQBtHpANpNeHbgC6AHTOqAKUEDo6cJSUfADWeiTiVja2UDGUXvEBJFDxMulOEMACCPgZsYVQAL58ljUZRPDIpljmSHoyac6l5cWuHjI1QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4AmsMolpFStYgENoUgLaJ8vdFF7QA9tEjgwAXxC6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA):

```ts
type ParsedOptions = Record<'a' | 'b', number>;

function parseOptions(): ParsedOptions {
  const result = ['a', 'b'].reduce((acc, k) => {
    acc[k] = k;
    return acc;
  }, {} as ParsedOptions);

  return result;
}
```

@bradzacher @JounQin what was the fix you had in mind? Adding an option to disable `as` casts?

Note that we're blocked from actually detecting unfixable reporting on my favorite TypeScript issue: https://github.com/microsoft/TypeScript/issues/9879. 😢

**Comment by bradzacher** [comment]
Potential fixes:
1) don't error if you use both:
```ts
  const result = ['a', 'b'].reduce<ParsedOptions>((acc, k) => {
    acc[k] = k;
    return acc;
  }, {} as ParsedOptions);
```
2) don't error if the value is specifically `{}` (maybe behind an option)

**Comment by bradzacher** [comment]
The big problem you've got is this code is straight-up unsound though - it's not good code!

For example Josh your code has `ParsedOptions` typed `Record<string, number>` - yet your code does `acc[k] = k` - creating an object like `Record<string, string>`!!! TS doesn't complain about it at all thanks to the completely unsafe `as`.

There's also the problem of missing keys - your type specifies the object must have the keys `'a' | 'b'`, but if you were to modify your array to just be `['a']` then the code would typecheck fine and TS would accept `result.b.toFixed()` - but it would crash at runtime because `b` isn't ever defined!

`.reduce` is just a bad way to construct an object with known keys, sadly.
And there's not really a good option otherwise because like `Object.fromEntries` forcibly returns `Record<string, T>` - so you can't type the keys.

**Comment by JoshuaKGoldberg** [comment]
Now that v8 is released, per https://github.com/typescript-eslint/typescript-eslint/discussions/7936, we can mark this issue as unblocked! 🚀
